File: FunctionalHttp.CSharpInterop/Core/HttpResponseExtensions.cs
using FunctionalHttp.Collections;
using Microsoft.FSharp.Collections;
using Microsoft.FSharp.Core;
using System;
using System.Collections.Generic;

using FunctionalHttp;

namespace FunctionalHttp.Core.Interop
{
    public static class HttpResponseExtensionsCSharp
    {
        public static HttpResponse<TResp> With<TResp>(
            this HttpResponse<TResp> This,
            AcceptableRanges acceptedRanges = null,
            TimeSpan? age = null,
            IEnumerable<Method> allowed = null,
            IEnumerable<Challenge> authenticate = null,
            IEnumerable<CacheDirective> cacheControl = null,
            ContentInfo contentInfo = null,
            DateTime? date = null,
            EntityTag etag = null,
            DateTime? expires = null,
            IEnumerable<Tuple<Header, object>> headers = null,
            Guid? id = null,
            DateTime? lastModified = null,
            Uri location = null,
            IEnumerable<Challenge> proxyAuthenticate = null,
            DateTime? retryAfter = null,
            Server server = null,
            Status status = null,
            Vary vary = null,
            HttpVersion? version = null,
            IEnumerable<Warning> warning = null)
        {
            return HttpResponseInternal.With<TResp, TResp>(
                acceptedRanges.ToFSharpOption(),
                age.ToFSharpOption(),
                allowed.ToFSharpOption(),
                authenticate.ToFSharpOption(),
                cacheControl.ToFSharpOption(), 
                contentInfo.ToFSharpOption(), 
                date.ToFSharpOption(),
                This.Entity,
                etag.ToFSharpOption(),
                expires.ToFSharpOption(), 
                headers.ToFSharpOption().Select(MapModule.OfSeq),
                id.ToFSharpOption(),
                lastModified.ToFSharpOption(),
                location.ToFSharpOption(), 
                proxyAuthenticate.ToFSharpOption(),
                retryAfter.ToFSharpOption(),
                server.ToFSharpOption(),
                status.ToFSharpOption(),
                vary.ToFSharpOption(),
                version.ToFSharpOption(),
                warning.ToFSharpOption(),
                This);
        }

        public static HttpResponse<TNew> With<TResp, TNew>(
            this HttpResponse<TResp> This,
            TNew entity,
            AcceptableRanges acceptedRanges = null,
            TimeSpan? age = null,
            IEnumerable<Method> allowed = null,
            IEnumerable<Challenge> authenticate = null,
            IEnumerable<CacheDirective> cacheControl = null,
            ContentInfo contentInfo = null,
            DateTime? date = null,
            EntityTag etag = null,
            DateTime? expires = null,
            IEnumerable<Tuple<Header, object>> headers = null,
            Guid? id = null,
            DateTime? lastModified = null, 
            Uri location = null,
            IEnumerable<Challenge> proxyAuthenticate = null,
            DateTime? retryAfter = null,
            Server server = null,
            Status status = null,
            Vary vary = null,
            HttpVersion? version = null,
            IEnumerable<Warning> warning = null)
        {
            return HttpResponseInternal.With<TResp, TNew>(
                acceptedRanges.ToFSharpOption(),
                age.ToFSharpOption(), 
                allowed.ToFSharpOption(),
                authenticate.ToFSharpOption(),
                cacheControl.ToFSharpOption(),
                contentInfo.ToFSharpOption(),
                date.ToFSharpOption(),
                entity,
                etag.ToFSharpOption(),
                expires.ToFSharpOption(),
                headers.ToFSharpOption().Select(MapModule.OfSeq),
                id.ToFSharpOption(),
                lastModified.ToFSharpOption(),
                location.ToFSharpOption(),
                proxyAuthenticate.ToFSharpOption(),
                retryAfter.ToFSharpOption(),
                server.ToFSharpOption(),
                status.ToFSharpOption(),
                vary.ToFSharpOption(),
                version.ToFSharpOption(),
                warning.ToFSharpOption(),
                This);
        }

        public static HttpResponse<TResp> Without<TResp>(
            this HttpResponse<TResp> This,
            bool acceptedRanges = false,
            bool age = false,
            bool allowed = false,
            bool authenticate = false,
            bool cacheControl = false,
            bool contentInfo = false,
            bool date = false,
            bool etag = false,
            bool expires = false,
            bool headers = false,
            bool lastModified = false,
            bool location = false,
            bool proxyAuthenticate = false,
            bool retryAfter = false,
            bool server = false,
            bool vary = false,
            bool warning = false)
        {
            return HttpResponseInternal.Without<TResp>(
                acceptedRanges,
                age, 
                allowed,
                authenticate, 
                cacheControl, 
                contentInfo, 
                date, 
                etag, 
                expires, 
                headers, 
                lastModified, 
                location, 
                proxyAuthenticate,
                retryAfter,
                server,
                vary,
                warning,
                This);
        }
    }
}

File: FunctionalHttp/Core.Interop/HttpResponseExtensions.fs
namespace FunctionalHttp.Core.Interop

open FunctionalHttp.Collections
open FunctionalHttp.Core
//open FunctionalHttp.Core.HttpResponseDeserializers
open System
open System.IO
open System.Runtime.CompilerServices

[<AbstractClass; Sealed; Extension>]
type HttpResponseExtensions private () =

    // FIXME: Prefer not to expose FSharpChoice in an interop API.
    [<Extension>]
    static member TryGetAcceptedRange(this:HttpResponse<'TResp>, acceptedRanges : byref<Choice<Set<RangeUnit>, AcceptsNone>>) = 
        Option.tryGetValue this.AcceptedRanges &acceptedRanges

    [<Extension>]
    static member TryGetAge(this:HttpResponse<'TResp>, age : byref<TimeSpan>) = 
        Option.tryGetValue this.Age &age

    [<Extension>]
    static member TryGetDate(this:HttpResponse<'TResp>, date : byref<DateTime>) = 
        Option.tryGetValue this.Date &date

    [<Extension>]
    static member TryGetETag(this:HttpResponse<'TResp>, etag : byref<EntityTag>) = 
        Option.tryGetValue this.ETag &etag

    [<Extension>]
    static member TryGetExpires(this:HttpResponse<'TResp>, expires : byref<DateTime>) = 
        Option.tryGetValue this.Expires &expires

    [<Extension>]
    static member TryGetLastModified(this:HttpResponse<'TResp>, lastModified : byref<DateTime>) = 
        Option.tryGetValue this.Expires &lastModified
         
    [<Extension>]
    static member TryGetLocation(this:HttpResponse<'TResp>, location : byref<Uri>) = 
        Option.tryGetValue this.Location &location

    [<Extension>]
    static member TryGetRetryAfter(this:HttpResponse<'TResp>, retryAfter : byref<DateTime>) = 
        Option.tryGetValue this.RetryAfter &retryAfter

    [<Extension>]
    static member TryGetServer(this:HttpResponse<'TResp>, server : byref<Server>) = 
        Option.tryGetValue this.Server &server
           
    [<Extension>]
    static member TryGetVary(this:HttpResponse<'TResp>, vary : byref<Choice<Set<Header>, Any>>) = 
        Option.tryGetValue (this.Vary) &vary       